cdap-security/src/main/java/io/cdap/cdap/security/authorization/InternalAccessEnforcer.java

```diff
@@ -0,0 +1,141 @@
+ * the License.
+ */
+
+package io.cdap.cdap.security.authorization;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.api.security.AccessException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.io.Codec;
+import io.cdap.cdap.proto.element.EntityType;
+import io.cdap.cdap.proto.id.EntityId;
+import io.cdap.cdap.proto.security.Credential;
+import io.cdap.cdap.proto.security.Permission;
+import io.cdap.cdap.proto.security.Principal;
+import io.cdap.cdap.security.auth.AccessToken;
+import io.cdap.cdap.security.auth.InvalidTokenException;
+import io.cdap.cdap.security.auth.TokenManager;
+import io.cdap.cdap.security.auth.UserIdentity;
+import io.cdap.cdap.security.auth.context.SystemAuthenticationContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.Set;
+
+public class InternalAccessEnforcer extends AbstractAccessEnforcer {
+  private static final Logger LOG = LoggerFactory.getLogger(InternalAccessEnforcer.class);
+
+  private final TokenManager tokenManager;
+  private final Codec<AccessToken> accessTokenCodec;
+
+  @Inject
+  InternalAccessEnforcer(CConfiguration cConf, TokenManager tokenManager, Codec<AccessToken> accessTokenCodec) {
+    super(cConf);
+    this.tokenManager = tokenManager;
+    this.accessTokenCodec = accessTokenCodec;
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Permission permission) throws AccessException {
+    doEnforce(entity, principal, Collections.singleton(permission));
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException {
+    doEnforce(entity, principal, permissions);
+  }
+
+  @Override
+  public void enforceOnParent(EntityType entityType, EntityId parentId, Principal principal, Permission permission)
+    throws AccessException {
+    doEnforce(entityType, parentId, principal, permission);
+  }
+
+  @Override
+  public Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal)
+    throws AccessException {
+    LOG.debug("Internal isVisible check for entities {} for principal {}", entityIds, principal);
+    try {
+      validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+      return entityIds;
+    } catch (AccessException e) {
+      return Collections.emptySet();
+    }
+  }
+
+  private void doEnforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException {
+    LOG.debug("Internal enforce check for entity {} for principal {} on permissions {}", entity, principal,
+              permissions);
+    validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+  }
+
+  private void doEnforce(EntityType entityType, EntityId parentId, Principal principal, Permission permission)
```

Comment:
Please name `doEnforceOnParent` to be clear

Reply:
Done